src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,181 @@
+ * and records their positions in the text.
+ *
+ * Author: Prabhat-Kumar-42
+ * GitHub: https://github.com/Prabhat-Kumar-42
+ */
+public class AhoCorasick {
+
+    // Trie Node Class
+    class Node {
+        HashMap<Character, Node> child = new HashMap<>();
+        Node suffix_link;
+        Node output_link;
+        int pattern_ind;
+
+        Node() {
+            this.suffix_link = null;
+            this.output_link = null;
+            this.pattern_ind = -1;
+        }
+    }
+
+    Node root = null; // The root node of the Aho-Corasick trie
+    private ArrayList<ArrayList<Integer>> res; // Stores the positions where patterns are found in the text
+    private String[] patterns; // patterns according to which Trie is constructed
+
+    // Clears the Aho-Corasick data structures
+    public void clear() {
+        root = null;
+        if (res != null) {
+            res.clear();
+        }
+    }
+
+    // Builds the Aho-Corasick trie from a set of input patterns
+    public void buildTrie(String[] patterns) {
+        root = new Node(); // Initialize the root of the trie
+        res = new ArrayList<>(patterns.length); // Initialize the result data structure
+        this.patterns = patterns;
+        // Loop through each input pattern
+        for (int i = 0; i < patterns.length; i++) {
+            res.add(new ArrayList<Integer>()); // Initialize a list to store positions of the current pattern
+
+            Node curr = root; // Start at the root of the trie for each pattern
+
+            // Loop through each character in the current pattern
+            for (int j = 0; j < patterns[i].length(); j++) {
+                char c = patterns[i].charAt(j); // Get the current character
+
+                // Check if the current node has a child for the current character
+                if (curr.child.containsKey(c)) {
+                    curr = curr.child.get(c); // Update the current node to the child node
+                } else {
+                    // If no child node exists, create a new one and add it to the current node's children
+                    Node nn = new Node();
+                    curr.child.put(c, nn);
+                    curr = nn; // Update the current node to the new child node
+                }
+            }
+            curr.pattern_ind = i; // Store the index of the pattern in the current leaf node
+        }
+    }
+
+    // Builds the suffix links and output links in the Aho-Corasick trie
+    public void buildSuffixAndOutputLinks() {
+        root.suffix_link = root; // Initialize the suffix link of the root to itself
+        Queue<Node> q = new LinkedList<>(); // Initialize a queue for BFS traversal
+
+        // Initialize suffix links for child nodes of the root
+        for (char rc : root.child.keySet()) {
+            q.add(root.child.get(rc)); // Add child node to the queue
+            root.child.get(rc).suffix_link = root; // Set suffix link to the root
+        }

```

Comment:
This could be moved to a separate method (called something like `initializeSuffixLinksForChildNodesOfTheRoot`). In general, a good method name is the best comment.

Reply:
I understand. Will Implement this.

Reply:
Added this method. It's also done now.